### PR TITLE
i'm getting really fucking tired of "edge case that when I test for it locally using the exact instructions from the players the crash doesn't occur" game

### PR DESCRIPTION
**Pull request by Iamgoofball**
:cl: Iamgoofball
add: Chemical Gas has been added to the game.
add: Add chems to a Chemical Gas Canister by hitting it with a beaker with the chemicals you want to add in it!
tweak: Freon is now colder, but no longer bypasses temperature protections.
/:cl:

where doin it

where makin it hapen

### CODE REVIEW

File: code/datums/gas_overrides.dm [hunk 1]
```diff
@@ -2,4 +2,7 @@
 	return 0
 
 /datum/proc/water_vapor_gas_act() // We get it
-	return 0 // You vape
\ No newline at end of file
+	return 0 // You vape
+
+/datum/proc/chem_gas_act(var/datum/reagents/gas)
```
Comment (RemieRichards) [inline code/datums/gas_overrides.dm]: you've been told about ```var/``` in proc arguments.
No more.

File: code/modules/atmospherics/environmental/LINDA_turf_tile.dm [hunk 2]
```diff
@@ -114,10 +114,17 @@
 /turf/open/proc/tile_graphic()
 	. = new /list
 	var/list/gases = air.gases
+	var/list/reagent_gas_memes = air.reagents.reagent_list
```
Comment (RemieRichards) [inline code/modules/atmospherics/environmental/LINDA_turf_tile.dm]: no.

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm [hunk 3]
```diff
@@ -353,6 +367,12 @@ var/list/gaslist_cache = null
 		removed.add_gas(id)
 		removed_gases[id][MOLES] = QUANTIZE((cached_gases[id][MOLES] / sum) * amount)
 		cached_gases[id][MOLES] -= removed_gases[id][MOLES]
+	for(var/R in cached_reagents)
+		var/datum/reagent/RA = R
+		var/temp_id = RA.id
+		var/mathmemes = QUANTIZE((RA.volume / reagents.total_volume) * amount)
```
Comment (RemieRichards) [inline code/modules/atmospherics/gasmixtures/gas_mixture.dm]: come on.

**Comment by bgobandit** [comment]
ok but what about reagent reactions in the air

**Review by RemieRichards — CHANGES_REQUESTED** [review]

**Comment by Iamgoofball** [comment]
they should react

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm (hunk 3 above)
Comment (Iamgoofball) [inline code/modules/atmospherics/gasmixtures/gas_mixture.dm]: what, it's temp vars and atmos memes

File: code/modules/atmospherics/environmental/LINDA_turf_tile.dm (hunk 2 above)
Comment (Iamgoofball) [inline code/modules/atmospherics/environmental/LINDA_turf_tile.dm]: it's for cacheing to improve speed

**Comment by PKPenguin321** [comment]
Truly airborne viruses was one of the most requested features when I first joined. Legend. :+1:

File: code/modules/reagents/chemistry/holder.dm [hunk 4]
```diff
@@ -61,163 +61,174 @@ var/const/INJECT = 5 //injection
...
+		cached_reagents.Cut()
+		cached_reagents = null
 	if(my_atom && my_atom.reagents == src)
 		my_atom.reagents = null
 
 /datum/reagents/proc/remove_any(amount = 1)
-	var/list/cached_reagents = reagent_list
-	var/total_transfered = 0
-	var/current_list_element = 1
+	if(reagent_list)
```
Comment (MrStonedOne) [inline code/modules/reagents/chemistry/holder.dm]: don't place an entire proc inside of an if. do `if(!reagent_list) return`

https://github.com/tgstation/tgstation/blob/master/CONTRIBUTING.md#use-early-return

File: code/modules/reagents/chemistry/holder.dm [hunk 5]
```diff
@@ -61,163 +61,174 @@ var/const/INJECT = 5 //injection
...
+		src.update_total()
+		R.update_total()
+		R.handle_reactions()
+		return amount
 
 /datum/reagents/proc/metabolize(mob/living/carbon/C, can_overdose = 0)
+	if(!reagent_list && !addiction_list)
+		reagent_list = list()
+		addiction_list = list()
 	var/list/cached_reagents = reagent_list
```
Comment (MrStonedOne) [inline code/modules/reagents/chemistry/holder.dm]: We have the lazy init list macros in helpers/list.dm for this.

Also, you should be checking lengths any time you remove stuff from a list and if it gets back to empty again, remove the list.

there is a macro in helpers/list.dm for that too.

File: code/modules/reagents/chemistry/holder.dm [hunk 6]
```diff
@@ -277,14 +288,15 @@ var/const/INJECT = 5 //injection
...
+		if(flags & REAGENT_NOREACT)
+			STOP_PROCESSING(SSobj, src)
+			return
 
-	for(var/reagent in cached_reagents)
-		var/datum/reagent/R = reagent
-		R.on_tick()
+		for(var/reagent in cached_reagents)
+			var/datum/reagent/R = reagent
+			R.on_tick()
```
Comment (MrStonedOne) [inline code/modules/reagents/chemistry/holder.dm]: this entire if isn't needed.

for treats null lists as if they were an empty list.

File: code/modules/reagents/chemistry/holder.dm (hunk 4 above)
Comment (Iamgoofball) [inline code/modules/reagents/chemistry/holder.dm]: Will do

File: code/modules/reagents/chemistry/holder.dm [hunk 7]
```diff
@@ -682,8 +714,9 @@ var/const/INJECT = 5 //injection
 	return trans_data
 
 /datum/reagents/proc/get_reagent(type)
-	var/list/cached_reagents = reagent_list
-	. = locate(type) in cached_reagents
+	if(reagent_list)
+		var/list/cached_reagents = reagent_list
+		. = locate(type) in cached_reagents
```
Comment (MrStonedOne) [inline code/modules/reagents/chemistry/holder.dm]: I'd test if this if is even needed, iirc locate(type) in list will treat a null list the same as an empty list.

**Comment by FantasticFwoosh** [comment]
Time to replace all the air with mercury or chlorine gas.

> Besides i have a especially evil gas in mind.

**Comment by Cebasiz** [comment]
we really need an industrial sized machine that takes massive amounts of power to synthesize enormous amounts of reagents so we can fill the whole air supply with wonderdrugs

thanks goof i expect it in a week

**Comment by BeeSting12** [comment]
so can i fill medbay with tricord and noone will OD on it?

**Comment by Iamgoofball** [comment]
@BeeSting12 nah, they'll OD if they hang in it too long, so be careful with that

**Comment by SgtxRootbeer** [comment]
Wait, what about miner's salve, does it cause cramps if taken this way? 
Because, one option is to flood the air supply with charcoal and miners salve instead of tricordazine

**Comment by TechnoAlchemisto** [comment]
How much gas would a 50u beaker release?

**Comment by Iamgoofball** [comment]
@TechnoAlchemisto It hangs out around 2.5u a tile, so it depends on how much chem gas you actually have,

**Comment by GunHog** [comment]
Is there a way to tell what reagents are contained within the gas?

**Comment by Iamgoofball** [comment]
breathe it in and scan yourself with a medical analyzer

**Comment by GunHog** [comment]
So, basically, "No". Should add it to air alarms and analyzers.

**Comment by FantasticFwoosh** [comment]
Just use your PDA air analyser (or the analyzer tool)

Scrubbers are issue number 2, could we have a generic (chem gas) scrubber active without listing off some odd 30+ options?

**Comment by DorfGuy1** [comment]
oh boy, this is going to be gr8

**Comment by KorPhaeron** [comment]
Didn't we turn this down already for being horrendously op and killing everyone every round

**Comment by Iamgoofball** [comment]
@KorPhaeron No, it got turned down due to issues with lists going batshit on the memory and killing the server from memory overuse

**Comment by Cebasiz** [comment]
it shouldn't be too bad, though you should definitely give the engineering sec outpost access to atmos on most maps because otherwise this is going to be a little crazy

**Comment by lordpidey** [comment]
Is there something to prevent using reagents that already have a gas?  It'd be kind of weird to have oxygen reagent gas.

File: code/modules/atmospherics/environmental/LINDA_turf_tile.dm [hunk 8]
```diff
@@ -114,10 +114,17 @@
...
 	. = new /list
 	var/list/gases = air.gases
+	var/list/reagent_gas = air.reagents.reagent_list
 	for(var/id in gases)
 		var/gas = gases[id]
 		if(gas[GAS_META][META_GAS_OVERLAY] && gas[MOLES] > gas[GAS_META][META_GAS_MOLES_VISIBLE])
-			. += gas[GAS_META][META_GAS_OVERLAY]
+			if(id == "chem_gas")
+				var/obj/effect/overlay/gas/G = gas[GAS_META][META_GAS_OVERLAY]
+				if(reagent_gas.len)
```
Comment (MrStonedOne) [inline code/modules/atmospherics/environmental/LINDA_turf_tile.dm]: needs a null check (or just use lazylen or just use length())

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm [hunk 9]
```diff
@@ -48,7 +49,8 @@ var/list/gaslist_cache = null
 	src.volume = volume
 	last_share = 0
 	fuel_burnt = 0
-
+	create_reagents(1000)
+	reagents.set_reacting(FALSE) // no
```
Comment (MrStonedOne) [inline code/modules/atmospherics/gasmixtures/gas_mixture.dm]: This is the kind of thing that might have to be lazyinit'ed also, it's a datum not a list so we don't have the macros for that, but every gas mixture doesn't need a reagent datum, only ones with reagents

Comment (Iamgoofball) [inline code/modules/atmospherics/gasmixtures/gas_mixture.dm]: Reagents datums aren't really supposed to stop existing after they start existing on an object. I can look into it, but no guarantees

Comment (optimumtact) [inline code/modules/atmospherics/gasmixtures/gas_mixture.dm]: they dont' even do anything goof, they just hold fukken reagents

**Comment by Shadowlight213** [comment]
@Iamgoofball does the admin message/logging actually WHAT chemicals are in the gas this time instead of "X has opened a canister containing the following: Chemicals"

**Comment by Iamgoofball** [comment]
@Shadowlight213 i guess I could but you can just jump to the tile and VV -> air -> reagents

**Comment by drynwyn** [comment]
Not if you're investigating something after the fact.

**Comment by Iamgoofball** [comment]
i guess
i'll look into it but if you just check it right off the bat you should be good
i can just make it log inserting chems into chem gas containers to save from opening-spam

**Comment by AdamElTablawy** [comment]
The possibility of bar neuro gas is a terrifying prospect. How will chemgas be available and what's the unit of reagent to gas content levels? 

Honestly, it might be better if smoke reactions made chemgas instead of having a snowflake canister available in weird spots because chemgas could very easily become much much worse than a plasmaflood (especially if some shitter release potassium and water gas. Explosions on every tile jfc)

**Comment by AdamElTablawy** [comment]
Also maybe analyzers should give feedback on what chems are in the gas?

I mean, initro gas or clf3 gas will be fairly obvious but lexorin chemgas might leave people confused.

Also does this apply in a touch, ingest, or inject reaction or are the chems just added

**Comment by Firecage** [comment]
Thanks for breaking the server goof.

**Comment by Iamgoofball** [comment]
i'm getting really fucking tired of "edge case that when I test for it locally using the exact instructions from the players the crash doesn't occur" game
